lightningd/pay.c

```diff
@@ -1496,6 +1521,77 @@ static const struct json_command listsendpays_command = {
 };
 AUTODATA(json_command, &listsendpays_command);
 
+
+static struct command_result *json_delpay(struct command *cmd,
+						const char *buffer,
+						const jsmntok_t *obj UNNEEDED,
+						const jsmntok_t *params)
+{
+	struct json_stream *response;
+	const struct wallet_payment **payments;
+	const char *status_str;
+	enum wallet_payment_status status;
+	struct sha256 *payment_hash;
+
+	if (!param(cmd, buffer, params,
+		p_req("payment_hash", param_sha256, &payment_hash),
+		p_opt("status", param_string, &status_str),
```

Comment:
Can use instead:

```c
p_opt_def("status", param_string, &status_str, "complete")
```

Reply:
@ZmnSCPxj, mh! maybe I'm losing something here, param_string accept only a `const char *str` and the function `o_opt_def` can not assign a new value at `status_str` variable.

With this comment, you mean that I need to declare some function that accepts a not const `char*`?

I think that I'm losing something, sorry about that

Reply:
String constants *are* `const char *`.  You can do something like `const char *foo = "foo";`.  You can even do something like `const char *foo = "foo"; foo = "bar"`.

The *correct* way to write `const` is *after* the type.  So actually `const char *` should be written as `char const *`.  Then you read types from right to left.  So `char const *` is read as "pointer (`*`) to constant (`const`) character (`char`)".  Note that the pointer itself is not `const`, so you can modify the pointer, which is why `char const* foo; foo = "bar"; foo = "quux"` is valid.  What you cannot do is `char const* foo = "foo"; *foo = '0';`, because what is pointed *to* is constant.

Reply:
Thanks for this wonderful informations. I made a couple of change inside https://github.com/ElementsProject/lightning/commit/627a374820dd03876eae9e4a99d661cb39f63934 but I want to have more detail about these changes. I like your suggestion to use p_opt_def.